controllers/user_controller.go

```diff
@@ -105,254 +106,381 @@ func (r *S3UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
-				fmt.Sprintf("The secret [%s] was created according to its matching custom resource", secret.Name), nil)
-		} else {
-			return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorFailed", metav1.ConditionFalse, "SecretCreationFailed",
-				fmt.Sprintf("A secret with the name %s already exist NOOP", secret.Name), nil)
+			// Add policies
+			err = r.addPoliciesToUser(ctx, userResource)
+			if err != nil {
+				return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorFailed", metav1.ConditionFalse, "S3UserPolicyAppendFailed",
+					fmt.Sprintf("Error while updating policies of user %s on S3 instance has failed", userResource.Name), err)
+			}
+
+			return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorSucceeded", metav1.ConditionTrue, "S3UserCreatedWithWithOverridenSecret",
+				fmt.Sprintf("The S3User %s was created and the Secret %s was updated successfully", userResource.Name, secret.Name), nil)
+		}
+
+		// Case 3.3 : they are not valid, and the operator is configured keep the existing secret
+		// The user will not be created, with no requeue and with two possible ways out : either toggle
+		// OverrideExistingSecret on, or delete the S3User whose credentials are not working anyway.
+		logger.Error(nil, fmt.Sprintf("A secret with the name %s already exists ; as the operator is configured to NOT override any pre-existing secrets, this user (%s) not be created on S3 backend until spec change (to target new secret), or until the operator configuration is changed to override existing secrets", secret.Name, userResource.Name))
+		return r.setS3UserStatusConditionAndUpdate(ctx, userResource, "OperatorSucceeded", metav1.ConditionTrue, "S3UserCreationFailedAsSecretCannotBeOverwritten",
+			fmt.Sprintf("The S3User %s wasn't created, as its spec targets a secret (%s) containing invalid credentials. The user's spec should be changed to target a different secret.", userResource.Name, secret.Name), nil)
+
+	}
+}
+
+func (r *S3UserReconciler) addPoliciesToUser(ctx context.Context, userResource *s3v1alpha1.S3User) error {
+	logger := log.FromContext(ctx).WithName("userCtrl")
+	policies := userResource.Spec.Policies
+	if policies != nil {
+		err := r.S3Client.AddPoliciesToUser(userResource.Spec.AccessKey, policies)
+		if err != nil {
+			logger.Error(err, "an error occurred while adding policy to user", "user", userResource.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *S3UserReconciler) handleS3UserDeletion(ctx context.Context, userResource *s3v1alpha1.S3User) (reconcile.Result, error) {
+	logger := log.FromContext(ctx).WithName("userCtrl")
+
+	if controllerutil.ContainsFinalizer(userResource, userFinalizer) {
+		// Run finalization logic for S3UserFinalizer. If the finalization logic fails, don't remove the finalizer so that we can retry during the next reconciliation.
+		if err := r.finalizeS3User(userResource); err != nil {
```

Comment:
name of the function not clear knowing that we delete S3user on minio  and not the crd here

Reply:
In our case, the operations needed to finalize the S3User resource is indeed to delete it from the S3 backend, but there could be potentially more. 

This "pattern" of delegating these operations in a dedicated method was taken straight from Operator SDK's doc : https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#handle-cleanup-on-deletion

And it's also used similarly for buckets, policies and paths. As such, I'd rather keep the function as is.